TestFunctional/parallel/TunnelCmd into serial subtest

govargo opened the pull request:
### What type of PR is this?
/kind cleanup
/area testing
/kind failing-test

### What this PR does / why we need it:

This is improvement & deep investigation PR for #7809.
Now, Hyperkit_macOS integration test always fails due to not be able to resolving DNS.
DNS resolution works in local macOS. So I think root cause of fail is releated to Jenkins environment or test server's environment.

This PR splits `TestFunctional/parallel/TunnelCmd` and adds DNS resolution test step.
And if DNS resolution test step is failed, we can get `dig` command output.
This will be helpful for deep investigation.

(EDIT)
This PR changed test `validateTunnelCmd` of `TestFunctional` in order to make sure test execution serially.

### Which issue(s) this PR fixes:

This will be helpful for #7809 and confirmed the issue fixed.
https://storage.googleapis.com/minikube-builds/logs/7956/884f0e2/HyperKit_macOS.html#pass_TestFunctional%2fparallel%2fTunnelCmd%2fserial%2fAccessThroughDNS

### Does this PR introduce a user-facing change?

No. This PR changes `TestFunctional/parallel/TunnelCmd` testcode.

**Before this PR**

Test `validateTunnelCmd` is tested in `TestFunctional`.

```
Test.1 access through loadbalancer ingress IP
Test.2 access through DNS resolution            ← Always fail, we cannot get more information
```

**After this PR**

`validateTunnelCmd` has serial subtests.

```
# TestTunnel(Serial Test)
Test.1 access through loadbalancer ingress IP
Test.2-1 DNS resolution (experimental) by dig
Test.2-2 DNS resolution (experimental) by dscacheutil
Test.3 access through DNS resolution
```

### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```
NONE
```

govargo commented:
/assign @priyawadhwa

medyagh commented:
/ok-to-test

govargo commented:
/retest

govargo commented:
I saw first time this failure...

```
E0502 02:40:48.608570   11443 out.go:102] [unset errFile]: * 
E0502 02:40:48.608798   11443 out.go:102] [unset errFile]: X Error getting primary control plane: remove /tmp/profile492025635/.minikube/profiles/config.json: no such file or directory
E0502 02:40:48.608826   11443 out.go:102] [unset errFile]: * 
E0502 02:40:48.608846   11443 out.go:102] [unset errFile]: * minikube is exiting due to an error. If the above message is not useful, open an issue:
E0502 02:40:48.608867   11443 out.go:102] [unset errFile]:   - https://github.com/kubernetes/minikube/issues/new/choose
FAIL	k8s.io/minikube/pkg/addons	0.032s
```

/retest

medyagh commented on test/integration/fn_tunnel_cmd.go:
```diff
@@ -136,12 +141,59 @@ func validateTunnelCmd(ctx context.Context, t *testing.T, profile string) {
+	}
+	dnsIP := "@" + ip.String()
+
+	// Test.2-1 DNS resolution by dig (experimental): https://minikube.sigs.k8s.io/docs/handbook/accessing/#dns-resolution-experimental
```
instead of adding comment test1
please use a t.Run("serial"){ } with 3 subtests, that run in a for loop. (to make sure they are in order. like this test
https://github.com/kubernetes/minikube/blob/100ba9b8d3464d22cffd9e02ab46b7a81c82d2ae/test/integration/pause_test.go#L38

govargo replied:
Thank for your advice!
I'll look it.

govargo commented:
@medyagh 
I tried `t.Run("serial"){ }` with subtests, but it doesn't work correclty first time.
Perhaps nested `t.Run` might have caused conflict.
(`TestFunctional(parallel)` -> `validateTunnelCmd` -> (added)`t.Run("serial"){}` not working )

So I made the tests independent such as `test/integration/pause_test.go`.
I renamed `test/integration/fn_tunnel_cmd.go` to `tunnel_test.go`.
And I deleted `validateTunnelCmd` from `TestFunctional`.

I don't understand E2E integration test scripts at all.
If I have to change those scripts, please let me know.

govargo commented:
/retest

govargo commented:
After trying source change, only `HyperKit_macOS` failed although nginx-svc has correct IP `10.108.42.59`.

```
--- FAIL: TestTunnel/serial/AccessDirect (135.43s)
tunnel_test.go:203: failed to hit nginx at "http://10.108.42.59": Temporary Error: Get http://10.108.42.59: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
tunnel_test.go:205: (dbg) Run:  kubectl --context tunnel-20200502040252-4586 get svc nginx-svc
tunnel_test.go:209: failed to kubectl get svc nginx-svc:
NAME        TYPE           CLUSTER-IP     EXTERNAL-IP    PORT(S)        AGE
nginx-svc   LoadBalancer   10.108.42.59   10.108.42.59   80:32189/TCP   2m25s
tunnel_test.go:216: expected body to contain "Welcome to nginx!", but got *""*
```

But other tests on other OS are success.

 * Docker_Linux  `TestTunnel/serial/AccessDirect` success
https://storage.googleapis.com/minikube-builds/logs/7956/83fd200/Docker_Linux.html#pass_TestTunnel%2fserial%2fAccessDirect

 * KVM_Linux `TestTunnel/serial/AccessDirect` success
https://storage.googleapis.com/minikube-builds/logs/7956/83fd200/KVM_Linux.html#pass_TestTunnel%2fserial%2fAccessDirect

Does anyone have any ideas why is macOS test failure?

govargo commented:
/retest

medyagh commented on test/integration/functional_test.go:
```diff
@@ -124,7 +124,6 @@ func TestFunctional(t *testing.T) {
 			{"ServiceCmd", validateServiceCmd},
 			{"AddonsCmd", validateAddonsCmd},
 			{"PersistentVolumeClaim", validatePersistentVolumeClaim},
-			{"TunnelCmd", validateTunnelCmd},
```
I think tunnel test should still be part of Functional Tests, since we consider this test as a Core Test that tests all functionalities of minikube.

when I meant to do it in serial, I meant tunnel test itself runs in parallel but all of it is own sub tests run in serial.

govargo replied:
@medyagh 
First time I tried the serial test into `validateTunnelCmd `, but the subtest failed because the process stopped like a deadlock.

After your this comment, I tried to add serial test into `validateTunnelCmd` of `TestFunctional` again.
And now serial test works and the `HyperKit_macOS` test all successed!!! 💯

https://storage.googleapis.com/minikube-builds/logs/7956/884f0e2/HyperKit_macOS.html#pass_TestFunctional%2fparallel%2fTunnelCmd%2fserial%2fAccessThroughDNS
